Update preview plan quiz

AlexVelezLl opened the pull request:
## Summary

Update quiz summary page to:
* Show learners and difficult questions report
* Show date created, date activated and class name in quiz status section
* Update the options button, and add a `preview` button in the header

Note that I havent updated page back buttons actions, as we will probably reestructure page routes in the future, and dont want to mess with it yet.

## References
Closes #12671.

## Screenshots
![image](https://github.com/user-attachments/assets/d80fe117-7d8e-4b53-9570-2ce3e3396e23)

![image](https://github.com/user-attachments/assets/523aa936-b385-484a-a157-ab4525b8cdbf)


## Reviewer guidance

Check that current quiz summary page match the acceptance criteria described in #12671.


----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

AlexVelezLl committed:
Move reports into quizsummarypage

AlexVelezLl committed:
Add ReportsDifficultQuestionsTable

AlexVelezLl committed:
add export csv

AlexVelezLl committed:
Add info to quizStatus

AlexVelezLl committed:
Add made visible label

AlexVelezLl committed:
Update preview button and quiz summary options

AlexVelezLl committed:
Update learners label

AlexVelezLl committed:
Fix print state

marcellamaki commented on kolibri/plugins/coach/assets/src/views/common/QuizLessonDetailsHeader.vue:
```diff
@@ -20,6 +20,7 @@
             />
           </h1>
           <StatusElapsedTime
+            v-if="examOrLesson !== 'exam'"
```
This is curiosity, not a leading question -- I can't quite figure out why this change is needed just from reading the diff. I can see that this component (the file `QuizLessonDetailsHeader` is used in other changed files, and that the `StatusElapsedTime` has also been updated. No need to add a code comment, but can you just explain to me? It's probably very straightforward but I just can't quite get there

AlexVelezLl replied:
Oh the thing is that we use this `QuizLessonDetailsHeader` for quizzes and lessons. And we were updating the quiz page to have the creation time as part of the `QuizStatus` component, so in the figma spec we have that we dont need to show the creation time here in the header anymore.

But as this header component was also being used for lessons, and we dont have yet the creation time in the `LessonStatus` component (which I assumed will be part of another issue to update the lessons page), I preferred to put this conditional of `v-if="examOrLesson !== 'exam'"` so that it is not shown in the quizzes as it is in the figma spec, but not "accidentally" remove it from the lessons yet, and rather do it on purpose once the lessons page has been refactored.

marcellamaki commented on kolibri/plugins/coach/assets/src/views/plan/QuizSummaryPage/index.vue:
```diff
@@ -122,36 +154,86 @@
+
+      // quizIsRandomized() {
+      //   return !this.quiz.learners_see_fixed_order;
+      // },
```
For this and for the "orderDescriptionString" - both have been commented out. Will we need them going forward?

AlexVelezLl replied:
Whoooppsss I forgot to remove them

marcellamaki commented on kolibri/plugins/coach/assets/src/routes/planExamRoutes.js:
```diff
@@ -53,8 +54,9 @@ export default [
-    path: '/:classId/plan/quizzes/:quizId',
+    path: '/:classId/plan/quizzes/:quizId/:tabId?',
     component: QuizSummaryPage,
+    handler: generateQuestionListHandler(['quizId']),
```
This handler, in a very not-straightforward way, is working in some cases, but causing a 500 error when you open a quiz that has not yet been started. Another function within the handler is dispatching `questionList/setItemStats`, which is assuming that `QuizDifficulties` (aka "Difficult Questions") exists. 

One way to go might be to updates the `setItemsStats` action, to me that seems as though it would be simpler to conditionalize that, rather that to recreate a similar action to build state for draft quizzes. It also seems to only be called in that one function (`showQuestionListView`), so I think there would be minimal chance of regression.

@rtibbles any thoughts here re: coach handlers/state management that I'm not thinking of?

rtibbles replied:
I guess the main thing is that for a draft quiz, we will get 500s because things are missing. We should update the API endpoint that is returning a 500, to instead return a 404 when the quiz is only a draft, and then make the handler handle the 404 gracefully (or just not even try it if it already knows the quiz is a draft).

More generally, most of the reports stuff is inapplicable when a quiz is still a draft, so we may need to utilize empty states here well too, and disabling tabs when not applicable?

marcellamaki replied:
From the code read through, I think there are already updates in the UI for empty state when not applicable, but the data is still being fetched (or, trying to be and erroring)

AlexVelezLl replied:
The problem is also with the handler, I havent looked in deph all that it was doing, but it was fetching de exam and the resources again, but we dont need this, as we already do this in the [fetchQuizSummaryPageData](https://github.com/AlexVelezLl/kolibri/blob/d915ea27c67992f6a9c290f618bffaa004f33e52/kolibri/plugins/coach/assets/src/views/plan/QuizSummaryPage/api.js#L4). So I will better remove this handler for this route and handle this within the fetchQuizSummaryPage function, and conditionally fetch the difficult questions if the exam is not a draft.

> so we may need to utilize empty states here well too, and disabling tabs when not applicable?

Yes, we currently have a condition to not show the tab if the quiz is a draft: https://github.com/AlexVelezLl/kolibri/blob/d915ea27c67992f6a9c290f618bffaa004f33e52/kolibri/plugins/coach/assets/src/views/plan/QuizSummaryPage/index.vue#L204

marcellamaki reviewed (changes requested):
Hi @AlexVelezLl - this looks great and is very nearly there! I did notice one problem with viewing "not started" quizzes due to some underlying assumptions in the "item stats" functions in the handlers. I think conditionalizing it might be the simplest path forward, and I added a few more details inline. 

It's okay for this to wait until after your trip to the mentor summit! It's not blocking other work that we will be working on next week.

AlexVelezLl committed:
Remove dead code

AlexVelezLl committed:
Fetch difficult questions in quizsummarypage api

AlexVelezLl commented:
Thank you @marcellamaki!! I have pushed the changes. At the end I fixed the 500 error in the api (just in case), and also removed the `generateQuestionListHandler` and handled the fetching of the difficult questions inside the quizSummaryPage api file: https://github.com/learningequality/kolibri/pull/12685/commits/a006f884eef5efc613be9d8bcefbd8a2dbaa2c5e

marcellamaki commented:
The code updates here make sense to me, and a quick manual testing checks out. I'm going to ask @pcenov to do some extra manual QA next week, but I think this is looking good!

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/QuizSummaryPage/api.js:
```diff
@@ -1,17 +1,52 @@
+  );
+
+  const allQuestions = exam.question_sources.reduce((qs, section) => {
+    qs = [...qs, ...section.questions];
+
+    return qs;
```
Non-blocking: You could  `return [...qs, ...section.questions]` here rather than assigning to the `qs` and returning it.

AlexVelezLl replied:
Oh yes, that would be much more simpler. Thanks!

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/QuizSummaryPage/index.vue:
```diff
@@ -255,6 +338,14 @@
+        });
+      },
+      exportCSV() {
+        this.$refs.table.exportCSV?.();
```
I'm not sure that the `?.` will avoid the error here. I tried this out in the node repl. I think that this is a blocker unless it's fully expected for `exportCSV` to not be defined, in which case this can be rewritten to avoid the resulting errors in the console.

```javascript
> x = {a: 1}
{ a: 1 }
> x.a?.()
Uncaught TypeError: x.a is not a function
```

AlexVelezLl replied:
Oh, no, no, the `?.()` is not supposed to be a validator to see if the `exportCSV` is a function, but rather that it is defined, and avoid breaking if its not. `exportCSV` is a very specific identifier and it is a verb, so having a non-function `exportCSV` attribute defined within the table ref component sounds like a very edge case. We could also write a couple of lines to make the validation validation more robust if needed, and just do
```javascript
if (typeof this.$refs.table.exportCSV === 'function') {
  this.$refs.table.exportCSV();
}
```

AlexVelezLl replied:
At the end, I just updated this, just in case, to make the typeof validation. :open_hands:

nucleogenesis replied:
Right on, thanks Alex. That makes sense and this way avoids the possible error altogether.

nucleogenesis reviewed (changes requested):
I may be mistaken or missing something re: my comment around the `?.()` but I'll consider it a blocking change until we can clarify.

pcenov commented:
LGTM as well @marcellamaki - no issues observed while manually testing other than the 'Preview' button not working and clicking the Back button which currently brings us to the Reports tab but both of these are out of scope for this PR.

AlexVelezLl committed:
Requested changes

AlexVelezLl commented:
Thank you @nucleogenesis! I just pushed the changes :)

nucleogenesis reviewed (approved):
Thanks Alex, the changes LGTM!

AlexVelezLl commented:
Thank you all!
